File: 1017.负二进制转换.go
/*
 * @lc app=leetcode.cn id=1017 lang=golang
 *
 * [1017] 负二进制转换
 */

// @lc code=start
func baseNeg2(n int) string {
	val := 0x55555555 ^ (0x55555555 - n)
	if val == 0 {
		return "0"
	}
	res := []byte{}
	for val > 0 {
		res = append(res, '0'+byte(val&1))
		val >>= 1
	}
	for i, n := 0, len(res); i < n/2; i++ {
		res[i], res[n-1-i] = res[n-1-i], res[i]
	}
	return string(res)
}

// @lc code=end

File: 1017.负二进制转换.cpp
/*
 * @lc app=leetcode.cn id=1017 lang=cpp
 *
 * [1017] 负二进制转换
 */
#include <string>
#include <iostream>
#include <vector>

using namespace std;

// @lc code=start
class Solution {
    public:
    string baseNeg2(int n)
    {
        if (n == 0)
        {
            return "0";
        }

        vector<int> bits(32);
        for (size_t i = 0; i < 32 && n != 0; i++)
        {
            if (n & 1)
            {
                bits[i]++;
                if (i & 1)
                {
                    bits[i + 1]++;
                }
            }
            n >>= 1;
        }
        int carry = 0;
        for (size_t i = 0; i < 32; i++)
        {
            int var = carry + bits[i];
            bits[i] = var & 1;
            carry = (var - bits[i]) / (-2);

        }
        int pos = 31;
        string result;
        while (pos >= 0 && bits[pos] == 0)
        {
            pos--;
        }

        while (pos >= 0)
        {
            result.push_back(bits[pos] + '0');
            pos--;
        }

        return result;
    }
};
// @lc code=end

